lib/src/code_field/search_result_highlighted_builder.dart

```diff
@@ -0,0 +1,165 @@
+import 'package:flutter/material.dart';
+
+import '../search/result.dart';
+import '../search/search_navigation_state.dart';
+
+@visibleForTesting
+const searchBackgroundColor = Color.fromARGB(141, 255, 235, 59);
+
+@visibleForTesting
+const searchTextColor = Colors.black;
+
+@visibleForTesting
+const currentMatchBackgroundColor = Colors.yellow;
+
+class SearchResultHighlightedBuilder {
+  final SearchResult searchResult;
+  final TextStyle? rootStyle;
+  final TextSpan textSpan;
+  late final int? highlightedMatchIndex;
+
+  SearchResultHighlightedBuilder({
+    required this.searchResult,
+    required this.rootStyle,
+    required this.textSpan,
+    required SearchNavigationState searchNavigationState,
+  }) {
+    if (searchResult.matches.isEmpty) {
+      return;
+    }
+
+    if (searchNavigationState.currentMatchIndex == null) {
+      highlightedMatchIndex = null;
+    } else {
+      highlightedMatchIndex = searchNavigationState.currentMatchIndex! * 2 + 1;
+    }
+
+    matchIndexes = searchResult.matches
+        .expand<int>((e) => [e.start, e.end])
+        .toList(growable: false);
+  }
+
+  /// List of spans to be added to the result.
+  final _spans = <InlineSpan>[];
+
+  /// Indexes of [searchResult] in ascending order.
+  /// Every element under even index is the start,
+  /// and every element under odd index is the end of the searchMatch.
+  ///
+  /// SearchMatches are grouped this way and look like this: 0-1 2-3 4-5
+  late final List<int> matchIndexes;
+
+  /// Current index of [matchIndexes] that is being processed.
+  int _currentMatchIndex = 0;
+
+  /// Whether the text before [_currentMatchIndex]
+  /// should have search or regular style.
+  bool get _isCurrentMatchIndexSearch => _currentMatchIndex.isOdd;
+
+  /// Whether we finished to process all of the search matches.
+  bool get _isLastMatchProcessed => _currentMatchIndex >= matchIndexes.length;
+
+  /// Number of characters that are already processed.
+  /// Or the current position in the text that we are about to process.
+  int _currentWindowStart = 0;
+
+  /// `TextStyle` of current span that is being processed.
+  TextStyle? _currentSpanStyle;
+
+  /// Returns the background color either for regular match,
+  /// or currently highlighted match.
+  Color get searchMatchBackgroundColor =>
+      highlightedMatchIndex == _currentMatchIndex
+          ? currentMatchBackgroundColor
+          : searchBackgroundColor;
+
+  /// Overrides `TextStyle` of span to highlight search result.
+  TextStyle get searchStyle =>
+      _currentSpanStyle?.copyWith(
+        backgroundColor: searchMatchBackgroundColor,
+        color: searchTextColor,
+      ) ??
+      TextStyle(
+        backgroundColor: searchMatchBackgroundColor,
+        color: searchTextColor,
+      );
+
+  /// Get actual style based on
+  /// whether the current processing index is inside search or not.
```

Comment:
/// The current style we are painting with:
/// either the span's original style, or the highlighted.